rfcs/0003-timestamps-and-ttl.md

```diff
@@ -0,0 +1,281 @@
+
+1. **Time-To-Live**: many applications, either for compliance or space-saving purposes
+   require purging data a certain amount of time after its insertion. Use cases with time
+   series data and privacy requirements such as GDPR come to mind here, though ensuring that 
+   deleted data is collected within a time bound is left for a future exercise.
+2. **Resolving Out-of-Order Operations**: some applications may process incoming requests
+   in a manner that results in out-of-order insertions to SlateDB (stream processing applications
+   are frequently subject to this type of pattern). Using insertion timestamps to resolve which
+   version of a key is more recent may be preferable to using the clock at the time of insertion.
+3. **Historical Reads / MVCC**: while SlateDB provides snapshot capabilities to ensure reads are
+   consistent and repeatable, it requires a snapshot to have been taken at the desired time. 
+   increasing the number of snapshots adds significant storage overhead. Instead, it is possible
+   to trade read-latency for space amplification by maintaining multiple versions of keys at
+   different timestamps to allow for similar "historical" read operations.
+
+## Goals
+
+- Define the API for specifying timestamps
+- Define the changes to the on-disk format of stored rows
+- Define the default semantics (and enumerate configurations) for out-of-order insertions
+- Outline mechanism to support Time-To-Live (TTL) via Compaction Filters
+
+## Out of Scope
+
+While the following are out of scope for the initial implementation, the design should
+not preclude:
+
+- Outlining compaction strategies that are optimized for TTL (such as Cassandra's
+  [TWCS](https://cassandra.apache.org/doc/stable/cassandra/operating/compaction/twcs.html)). This
+  will require additional metadata per-SST and a compaction strategy to match, but is otherwise
+  possible with what's outlined in the scope of this design.
+- Outlining compaction strategies that target efficient deletes (such as outlined
+  in [Lethe](https://disc-projects.bu.edu/lethe/))
+- Ensuring deletions are resolved within a time bound (a TTL does not ensure the data is deleted
+  from the object store within the specified time)
+- A public API for compaction filters (see https://github.com/slatedb/slatedb/issues/225) - this RFC
+  will implement TTL using a compaction filter but will not yet expose a public API for plugging in
+  custom compaction filters. Instead, the implementation should ensure that such an extension is
+  easy.
+
+## Public API
+
+### Configuration
+
+We will make the following changes to `DbOptions`:
+- introduce a configuration for a custom clock, and will default to using the system clock
+- introduce a configuration for a default TTL, which defaults to -1
+
+```rust
+/// defines the clock that SlateDB will use during this session
+pub trait Clock {
+    fn now(&self) -> i64;
```

Comment:
does `i64` make sense here? not sure what is common for interoperability with other Rust libraries (this is what ChatGPT suggested)

Reply:
Clarified that it's expected to be millis, though I guess that's somewhat arbitrary. It just has to match the unit you use for ttl and be monotonically increasing.

Reply:
Yea. Technically it doesn't even have to be time. It could just as easily be Kafka offsets or an auto-increment ID or anything else...

Reply:
Maybe we should document it as such.. it's just some kind of logical or physical clock...